<H1> ACTIVITY #2 :UK ACCIDENTS </H1>
<hr>
<h3> Analysts : Quert Russel Lalisan</h3>

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import f_oneway

import warnings
warnings.filterwarnings("ignore")

In [2]:
accident = pd.read_csv('dataset\\uk_road_accident.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [5]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [6]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mean())
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mean())
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mode()[0])

<h1>1) What is the overall distribution of Accident Severity (e.g., Fatal, Serious, Slight)?</h1>

In [7]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h2>Insights : The overall distribution of Accidents severity are </h2>
<h4>from the lowest being Fatal with 8661 recorded accidents<br>from the middle range being Serious with 88217 recorded accidents<br>from the lowest being Fatal with 563801 recorded accidents</h4>

<h1>2) Which District Area reports the highest number of severe accidents?</h1>

In [8]:
accident.groupby('Accident_Severity')['District Area'].value_counts()

Accident_Severity  District Area           
Fatal              Birmingham                  105
                   Leeds                        93
                   Highland                     88
                   East Riding of Yorkshire     85
                   Bradford                     71
                                              ... 
Slight             Berwick-upon-Tweed          121
                   Shetland Islands            113
                   Teesdale                    110
                   Orkney Islands               96
                   Clackmannanshire             72
Name: count, Length: 1266, dtype: int64

In [9]:
accident[accident['Accident_Severity'] == 'Fatal']['District Area'].describe()

count           8661
unique           422
top       Birmingham
freq             105
Name: District Area, dtype: object

In [10]:
accident[accident['Accident_Severity'] == 'Serious']['District Area'].describe()

count          88217
unique           422
top       Birmingham
freq            1474
Name: District Area, dtype: object

<h2>Insights : The District Area with highest Accident Severity is <b>"Birmingham"</b> </h2>
<h4> With a toal value of <b> 105 Fatal accidents </b> only in Birmingham <br> and a total value of <b> 1474 Fatal accidents  </b>in Birmingham </h4>

<h1>3) How does the number of casualties differ by Accident Severity?</h1>

In [11]:
accident.groupby('Accident_Severity')['Number_of_Casualties'].describe()

count      mean       std  min  25%  50%  75%   max
Accident_Severity                                                        
Fatal                8661.0  1.903129  1.901614  1.0  1.0  1.0  2.0  68.0
Serious             88217.0  1.467280  1.027915  1.0  1.0  1.0  2.0  45.0
Slight             563801.0  1.331402  0.754502  1.0  1.0  1.0  1.0  47.0

In [12]:
accident.groupby('Accident_Severity')['Number_of_Casualties'].mean()

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h2> Insights : The number of Casualties by Accident severity (Serious, Slight, Fatal) </h2>
<h4 style=textcolor:"blue"> Most accidents across all severity levels involve only one casualty, but Fatal accidents tend to have more people affected on average (≈1.9 per accident) compared to Serious (≈1.5) and Slight (≈1.3). This suggests that when an accident is fatal, it is more likely to involve multiple casualties, and although rare, some accidents can result in extremely high casualty counts (up to 68).</h4>

<h1>4) On average, how many vehicles are involved in accidents by severity level?</h1>

In [13]:
accident.groupby('Accident_Severity')['Number_of_Vehicles'].describe()

count      mean       std  min  25%  50%  75%   max
Accident_Severity                                                        
Fatal                8661.0  1.786976  1.022719  1.0  1.0  2.0  2.0  28.0
Serious             88217.0  1.678327  0.751475  1.0  1.0  2.0  2.0  19.0
Slight             563801.0  1.855864  0.700571  1.0  1.0  2.0  2.0  32.0

In [14]:
accident.groupby('Accident_Severity')['Number_of_Vehicles'].mean()

Accident_Severity
Fatal      1.786976
Serious    1.678327
Slight     1.855864
Name: Number_of_Vehicles, dtype: float64

<h4> Insights : for every severity level there is an average of <b>1.786976 vehicles</b> for cases that has severity of fatal, <b>1.678327 vehicles</b> for cases that are serious, whilst on Slight <b> 1.855864 vehicles</b> involve in accidents</h4>

<h1>5) Which dates had the highest accident counts?</h1>

In [15]:
accident['Accident Date'].value_counts()

Accident Date
30-11-2019    704
31-01-2019    697
13-07-2019    692
13-11-2021    692
14-08-2019    688
             ... 
30-12-2022    171
25-12-2019    157
25-12-2022    145
10/1/2022     123
25-12-2020    118
Name: count, Length: 1461, dtype: int64

<h2> Insights : The top 3 dates with the highest accidents occured are the following  </h2>
<h4> <b>- November 11, 2019 </b> with <b> 704 accidents </b> that happened <br>  <b>- January 1, 2019 </b> with <b> 697 accidents </b> that occured<br> <b>- July 13, 2019 </b> with <b> 692 accidents </b> that occured</h4>

<h1>6) Is accident severity correlated with the number of casualties??</h1>

In [16]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660679.000000  660679.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406896       1.383302              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490696      -2.332191              1.000000            1.000000
50%        52.315755      -1.411761              1.000000            2.000000
75%        53.453444      -0.232901              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [17]:
groups = [
    grp["Number_of_Casualties"].values
    for name, grp in accident.groupby("Accident_Severity")
]

f_stat, p_val = f_oneway(*groups)

print("F-statistic:", f_stat)
print("p-value:", p_val)

F-statistic: 2985.01643111389
p-value: 0.0


<h3> Insights : There are no correlation between Severity of accidents to the number of casualties</h3>

<h1>7) Do accidents cluster more in urban or rural areas?</h1>

In [18]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h4> Insights : According to the Data accidents usually happens on Urban areas with 421,678 cases of accidents and 238,990 cases on Rural areas </h4>

<h1>8) Is there a correlation between the number of vehicles involved and the number of casualties?</h1>

In [19]:
accident['Number_of_Vehicles'].corr(accident['Number_of_Casualties'])

np.float64(0.22888886126927557)

<h4> Insights : There is a <b> Weak positive correlation (0.22) </b> between The number of Vehicles involved and the number of casualties as the results showed </h4>

<h1>9) How do Weather Conditions impact accident severity?</h1>

In [20]:
accident.groupby('Accident_Severity')['Weather_Conditions'].describe()

count unique                 top    freq
Accident_Severity                                           
Fatal                8661      8  Fine no high winds    7207
Serious             88217      8  Fine no high winds   73285
Slight             563801      8  Fine no high winds  454521

<h4> Insights : The weather doesn't seem to impact the Severity of accidents, as most accidents happens during <b> Fine no high winds </b> rather than what was thought to be rainy days</h4>

<h1>10) What is the relationship between Light Conditions (daylight, darkness, streetlights) and accidents?</h1>

In [21]:
groups = [
    grp["Number_of_Casualties"].values
    for name, grp in accident.groupby("Light_Conditions")
]

f_stat, p_val = f_oneway(*groups)

print("F-statistic:", f_stat)
print("p-value:", p_val)

F-statistic: 534.4033532420921
p-value: 0.0


In [22]:
accident['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h3> Insights : According to the data There is a lot more accidents that happens on Broad Daylight as it was during night hours when there are no Lights</h3>

<h1>11) Do Road Surface Conditions (dry, wet, icy) affect the number of casualties?</h1>

In [23]:
accident.groupby('Road_Surface_Conditions')['Number_of_Casualties'].describe()

count      mean       std  min  25%  50%  75%  \
Road_Surface_Conditions                                                     
Dry                      448547.0  1.334545  0.794438  1.0  1.0  1.0  1.0   
Flood over 3cm. deep       1017.0  1.495575  0.977216  1.0  1.0  1.0  2.0   
Frost or ice              18517.0  1.348815  0.838040  1.0  1.0  1.0  1.0   
Snow                       5890.0  1.359593  0.765221  1.0  1.0  1.0  2.0   
Wet or damp              186708.0  1.411064  0.890931  1.0  1.0  1.0  2.0   

                          max  
Road_Surface_Conditions        
Dry                      45.0  
Flood over 3cm. deep      9.0  
Frost or ice             48.0  
Snow                     18.0  
Wet or damp              68.0

<h3> Insights : Road surface conditions does affect the number of Casualties, Having dry roads having the most amount of accidents occured compare to Wet,damp or snowy roads</h3>

<h1>12) Which Road Type (single carriageway, dual carriageway, roundabout, etc.) has the most accidents?</h1>

In [24]:
accident['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h3> Insigths : The type of road that has the most frequent accidents are <b> Single Carriageway </b> with a total number accidents of 496,663 </h3>

<h1>13) Which Vehicle Type is most frequently involved in accidents?</h1>

In [25]:
accident['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h3> Insights : Vehicles such as Cars, Van and Bus are the top 3 most frequently used vehicles that are involved in accidents. Car having <b> 497,992 involved accidents </b>, Van having <b> 34,160 involved accidents </b>, and Bus having<b> 25,878 involved accidents </b></h3>

<h1>14) Do motorcycles vs. cars show higher casualty counts?</h1>

In [26]:
accident[accident['Vehicle_Type'] == 'Car'] .describe().sum()

Latitude                498313.885956
Longitude               497982.198341
Number_of_Casualties    498066.184790
Number_of_Vehicles      498032.548149
dtype: float64

In [27]:
accident[accident['Vehicle_Type'] == 'Motorcycle over 500cc'].describe().sum()

Latitude                25978.379555
Longitude               25647.594423
Number_of_Casualties    25690.179867
Number_of_Vehicles      25676.543977
dtype: float64

<h3> Insights : According to the data Cars shows a higher casualty counts having <b> 498,066 </b> of casualties, over Motorcylces with <b> 25,690 </b> number of casualties </h3>

<h1>15) What is the relationship between Number of Vehicles and Number of Casualties in an accident?</h1>

In [28]:
accident['Number_of_Vehicles'].corr(accident['Number_of_Casualties'])

np.float64(0.22888886126927557)

In [40]:
accident.groupby('Number_of_Vehicles')['Number_of_Casualties'].count()

Number_of_Vehicles
1     200787
2     391995
3      52806
4      11300
5       2464
6        758
7        296
8        147
9         57
10        27
11        12
12         7
13         6
14         9
15         1
16         4
19         1
28         1
32         1
Name: Number_of_Casualties, dtype: int64

<h3> Insights : There is a moderate positive correlation (r = 0.23) between the number of vehicles involved and the number of casualties. On average, single-vehicles accidents tend to result in more casualties than single-vehicle accidents.</h3>

<h1>16) Are there particular District Areas where heavy vehicles cause more severe accidents?</h1>

In [42]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660679 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660679 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  660679 non-null  object 
 10  Road_Type                660679 non-null  object 
 11  Urban_or_Rural_Area      660679 non-null  object 
 12  Weather_Conditions       660679 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [75]:
heavy_vehicles = [
    'Bus or coach (17 or more pass seats)',
    'Van / Goods 3.5 tonnes mgw or under',
    'Motorcycle over 500cc',
    'Goods over 3.5t. and under 7.5t',
    'Agricultural vehicle',
    'Minibus (8 - 16 passenger seats)'
]

ays = accident[
    (accident['Vehicle_Type'].isin(heavy_vehicles)) &
    (accident['Accident_Severity'].isin(['Fatal', 'Serious']))
][['District Area', 'Accident_Severity']]

# Count combinations
ays.value_counts()

District Area    Accident_Severity
Birmingham       Serious              257
Westminster      Serious              171
Leeds            Serious              161
Sheffield        Serious              128
Bradford         Serious              107
                                     ... 
Mid Devon        Fatal                  1
Medway           Fatal                  1
Mansfield        Fatal                  1
Castle Point     Fatal                  1
West Lancashire  Fatal                  1
Name: count, Length: 794, dtype: int64

<h3> Insights : The districts area that has large amount of severe accidents (Fatal and Serious) involving heavy vehicles such as bus etc. are Birmingham District Area with 257 Serious Accidents and Westminster with 171 </h3>

<h1>17) Compare Urban vs. Rural accident severity distributions.</h1>

In [29]:
accident.groupby('Accident_Severity')['Urban_or_Rural_Area'].describe()

count unique    top    freq
Accident_Severity                              
Fatal                8661      2  Rural    5601
Serious             88217      3  Urban   50904
Slight             563801      3  Urban  367714

<h3> Insights : According to the data Rural places seems to Have the most Fatak accidents, whereas Urban areas have the most serious and slight recorded cases of accidents </h3>

<h1>18) Do weather conditions combined with light conditions (e.g., rain at night) show higher accident severity?</h1>

In [80]:
accident.groupby(['Light_Conditions', 'Weather_Conditions'])['Accident_Severity'].value_counts()

Light_Conditions             Weather_Conditions     Accident_Severity
Darkness - lighting unknown  Fine + high winds      Slight                 49
                                                    Serious                 7
                                                    Fatal                   3
                             Fine no high winds     Slight               4614
                                                    Serious               662
                                                                         ... 
Daylight                     Snowing + high winds   Serious                50
                                                    Fatal                   1
                             Snowing no high winds  Slight               3163
                                                    Serious               292
                                                    Fatal                  20
Name: count, Length: 113, dtype: int64

In [85]:
accident.groupby(['Light_Conditions', 'Weather_Conditions'])['Accident_Severity'].describe()

count unique     top  \
Light_Conditions            Weather_Conditions                             
Darkness - lighting unknown Fine + high winds          59      3  Slight   
                            Fine no high winds       5333      3  Slight   
                            Fog or mist                65      2  Slight   
                            Other                     186      3  Slight   
                            Raining + high winds      111      3  Slight   
                            Raining no high winds     621      3  Slight   
                            Snowing + high winds       15      1  Slight   
                            Snowing no high winds      94      2  Slight   
Darkness - lights lit       Fine + high winds        1852      3  Slight   
                            Fine no high winds      93958      3  Slight   
                            Fog or mist               921      3  Slight   
                            Other                    4803      3  Slight   
                            Raining + high winds     3066      3  Slight   
                            Raining no high winds   22666      3  Slight   
                            Snowing + high winds      241      3  Slight   
                            Snowing no high winds    1828      3  Slight   
Darkness - lights unlit     Fine + high winds          51      3  Slight   
                            Fine no high winds       1745      3  Slight   
                            Fog or mist                37      3  Slight   
                            Other                     143      2  Slight   
                            Raining + high winds       73      3  Slight   
                            Raining no high winds     459      3  Slight   
                            Snowing + high winds        5      2  Slight   
                            Snowing no high winds      30      2  Slight   
Darkness - no lighting      Fine + high winds         796      3  Slight   
                            Fine no high winds      25251      3  Slight   
                            Fog or mist               862      3  Slight   
                            Other                    1915      3  Slight   
                            Raining + high winds     1423      3  Slight   
                            Raining no high winds    6208      3  Slight   
                            Snowing + high winds      171      3  Slight   
                            Snowing no high winds     811      3  Slight   
Daylight                    Fine + high winds        5796      3  Slight   
                            Fine no high winds     408726      3  Slight   
                            Fog or mist              1643      3  Slight   
                            Other                   10103      3  Slight   
                            Raining + high winds     4942      3  Slight   
                            Raining no high winds   49742      3  Slight   
                            Snowing + high winds      453      3  Slight   
                            Snowing no high winds    3475      3  Slight   

                                                     freq  
Light_Conditions            Weather_Conditions             
Darkness - lighting unknown Fine + high winds          49  
                            Fine no high winds       4614  
                            Fog or mist                59  
                            Other                     154  
                            Raining + high winds       95  
                            Raining no high winds     550  
                            Snowing + high winds       15  
                            Snowing no high winds      86  
Darkness - lights lit       Fine + high winds        1497  
                            Fine no high winds      77944  
                            Fog or mist               773  
                            Other                    4213  
                   

<h3>Insigths : Weather conditions combined with poor lighting, such as rain, fog, or snow at night, lead to disproportionately higher accident severity, with darkness and lack of road lighting emerging as critical risk factors compared to similar conditions in daylight. </h3>

<h1>19) What trends in Accident Severity are visible over time (e.g., decreasing or increasing)?</h1>

In [91]:
accident.groupby('Accident Date')['Accident_Severity'].value_counts()

Accident Date  Accident_Severity
1/1/2019       Slight               271
               Serious               53
               Fatal                 11
1/1/2020       Slight               201
               Serious               49
                                   ... 
9/9/2021       Serious               86
               Fatal                  6
9/9/2022       Slight               381
               Serious               71
               Fatal                  8
Name: count, Length: 4370, dtype: int64

<h3> Insights : The </h3>

<h1>20) Can we identify the top 5 most dangerous accident scenarios (combination of road type, weather, light, vehicle type)?</h1>

In [93]:
accident.groupby(['Light_Conditions', 'Weather_Conditions','Vehicle_Type','Road_Type'])['Accident_Severity'=='Fatal'].describe()

KeyError: 'Column not found: False'